meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClientAsyncCallback.java

```diff
@@ -0,0 +1,124 @@
+ * thread. In our first version of implementation, we will keep similar behavior and have
+ * callbacks executed in ZkClient event thread, and reuse zkclient retry logic.
+ */
+public class ZkMetaClientAsyncCallback extends ZkAsyncCallbacks {
+
+  public static class ZkMetaClientCreateCallbackHandler extends CreateCallbackHandler {
+    AsyncCallback.VoidCallback _userCallback;
+    Object _userContext;
+
+    public ZkMetaClientCreateCallbackHandler(AsyncCallback.VoidCallback cb, Object context) {
+      _userCallback = cb;
+      _userContext = context;
+    }
+
+    @Override
+    public void handle() {
+      _userCallback.processResult(getRc(), getPath(), _userContext);
+    }
+
+  }
+
+  public static class ZkMetaClientDeleteCallbackHandler extends DeleteCallbackHandler {
+    AsyncCallback.VoidCallback _userCallback;
+    Object _userContext;
+
+    public ZkMetaClientDeleteCallbackHandler(AsyncCallback.VoidCallback cb, Object context) {
+      _userCallback = cb;
+      _userContext = context;
+    }
+
+    @Override
+    public void handle() {
+      _userCallback.processResult(getRc(), getPath(), _userContext);
+    }
+  }
+
+  public static class ZkMetaClientSetCallbackHandler extends SetDataCallbackHandler {
+    AsyncCallback.StatCallback _userCallback;
+    Object _userContext;
+
+    public ZkMetaClientSetCallbackHandler(AsyncCallback.StatCallback cb, Object context) {
+      _userCallback = cb;
+      _userContext = context;
+    }
+
+    @Override
+    public void handle() {
+      _userCallback.processResult(getRc(), getPath(), _userContext, getStat() == null ? null
+          : new MetaClientInterface.Stat(
+              ZkMetaClientUtil.convertZkEntryModeToMetaClientEntryMode(getStat().getEphemeralOwner()),
+              getStat().getVersion()));
+    }
+  }
+
+  public static class ZkMetaClientExistCallbackHandler extends ExistsCallbackHandler {
+    AsyncCallback.StatCallback _userCallback;
+    Object _userContext;
+
+    public ZkMetaClientExistCallbackHandler(AsyncCallback.StatCallback cb, Object context) {
+      _userCallback = cb;
+      _userContext = context;
+    }
+
+    @Override
+    public void handle() {
+      _userCallback.processResult(getRc(), getPath(), _userContext, getStat() == null ? null
+          : new MetaClientInterface.Stat(
+              ZkMetaClientUtil.convertZkEntryModeToMetaClientEntryMode(getStat().getEphemeralOwner()),
+              getStat().getVersion()));
+    }
+  }
+
+  public static class ZkMetaClientGetCallbackHandler extends GetDataCallbackHandler {
+    AsyncCallback.DataCallback _userCallback;
+    Object _userContext;
+
+    public ZkMetaClientGetCallbackHandler(AsyncCallback.DataCallback cb, Object context) {
+      _userCallback = cb;
+      _userContext = context;
+    }
+
+    @Override
+    public void handle() {
+      _userCallback.processResult(getRc(), getPath(), _userContext, getData(),
+          getStat() == null ? null : new MetaClientInterface.Stat(
```

Comment:
same here, why we create the stats but not modify the ZNode.

Reply:
Resolved per our offline discussion.